### PR TITLE
Block Function

### LINKED ISSUE #1009: Restore the ability to "Block" (defend against attacks).
**Is your feature request related to a problem? Please describe.**
The original concept of the block is very interesting (video below), it would be interesting to bring it back. There is mention of the block in several places, server configuration, action button when equipped with a shield and so on.

**Describe the solution you'd like**
New features can be added to this feature:
In the item editor, when we select shield, it's like normal equipment, there are no differences, for it in the item editor could have

1 - Success Chance % - A percentage of successfully defending the damage/projectile, so for each hit/projectile deferred, pressing the protection button would give you x% chance to successfully defend. (a different chance for each shield, defined by the creator, in the item editor)

2 - Damage Absorption % - A percentage of the damage that will be absorbed when the user succeeds in a defensive attempt.
(In entity.cs -> Attack(), before damageBase != 0, the calculation of the damage that will be subtracted from the victim's health is done, after this calculation it could enter the calculation of the absorption when it is a success, so if the damage continues other than 0 continues, if not, the damage is null.) (that's a big "I think", I'm not the best person to know how to implement)

3 - To do something to damage spells maybe, or can the shield work for all types of damage? (Spell, projectile, armed/unarmed combat)

4 - To do something for mana damage maybe

**Additional context**
https://www.youtube.com/watch?v=TgmhdnsR6TM&ab_channel=AscensionGameDev

### PR DESCRIPTION
**Pull request by WeylonSantana**
Resolves #1009

Block worked a few years ago, [this was its original function](https://www.youtube.com/watch?v=TgmhdnsR6TM&ab_channel=AscensionGameDev). But since it stopped working, traces of its code continue to this day. (Blocking properties on the entity, configuration on the server, shortcut button on the client, block action messages, etc.). I decided to repurpose this and rethink the way block works

I will be using these 3 properties
image

Block Chance: Chance the user has to successfully block a hit.
Block Amount: When successfully defending, percentage of damage that the block action will decrease.
Block Damage Absorption: When successfully defending and the damage absorbed, a percentage of that damage that will increase the health of the defending user (a kind of reverse lifesteal)

Simple test (we can consider armed or unarmed combat - Absorption not working yet): https://www.ascensiongamedev.com/resources/filehost/6ea7556f566d4ce2f1a7f1606c9ed09b.mp4

### CODE REVIEW

**Comment by WeylonSantana** [comment]
Using the damage absorbed by the shield to increase health is working:
https://www.ascensiongamedev.com/resources/filehost/f597ab556846c4de00daf65318614835.mp4

File: Intersect.Client/Interface/Game/DescriptionWindows/ItemDescriptionWindow.cs [hunk 1]
```diff
@@ -265,6 +265,27 @@ protected void SetupEquipmentInfo()
                     // Percentage based.
                     rows.AddKeyValueRow(Strings.ItemDescription.AttackSpeed, Strings.ItemDescription.Percentage.ToString(mItem.AttackSpeedValue));
                 }
+
+            }
+
```
Comment (Arufonsu) [inline Intersect.Client/Interface/Game/DescriptionWindows/ItemDescriptionWindow.cs:270]: Unnecessary extra spacing.

File: Intersect.Server/Entities/Entity.cs [hunk 2]
```diff
@@ -1753,6 +1753,54 @@ public virtual void TryAttack(Entity target)
             );
             }
 
+            //Check on each attack if the enemy is a player AND if he is blocking.
```
Comment (Arufonsu) [inline Intersect.Server/Entities/Entity.cs:1756]: > //Check on each attack if the enemy is a player AND if he is blocking.

* use singular third-person pronoun:
_//Check on each attack if the enemy is a player AND if **they** are blocking._

source: https://apastyle.apa.org/blog/singular-they

File: Intersect.Server/Entities/Entity.cs [hunk 3]
```diff
@@ -1753,6 +1753,54 @@ public virtual void TryAttack(Entity target)
...
+            if (enemy is Player player && player.Blocking)
+            {
+                //Getting the ID and Item the player is currently using.
+                var itemId = player.Items[player.Equipment[Options.ShieldIndex]].ItemId;
+                var item = ItemBase.Get(itemId);
+
+                if (item != null)
+                {
+                    int originalBaseDamage = baseDamage;
```
Comment (Arufonsu) [inline Intersect.Server/Entities/Entity.cs:1765]: use `var`

File: Intersect.Server/Entities/Entity.cs [hunk 4]
```diff
@@ -1753,6 +1753,54 @@ public virtual void TryAttack(Entity target)
...
+                    int blockChance = item.BlockChance;
```
Comment (Arufonsu) [inline Intersect.Server/Entities/Entity.cs:1766]: use `var`

File: Intersect.Server/Entities/Entity.cs [hunk 5]
```diff
@@ -1753,6 +1753,54 @@ public virtual void TryAttack(Entity target)
...
+                    double blockAmount = item.BlockAmount / 100.0;
```
Comment (Arufonsu) [inline Intersect.Server/Entities/Entity.cs:1767]: use `var`

File: Intersect.Server/Entities/Entity.cs [hunk 6]
```diff
@@ -1753,6 +1753,54 @@ public virtual void TryAttack(Entity target)
...
+                    double blockAbsorption = item.BlockAbsorption / 100.0;
```
Comment (Arufonsu) [inline Intersect.Server/Entities/Entity.cs:1768]: use `var`

File: Intersect.Server/Entities/Entity.cs [hunk 7]
```diff
@@ -1753,6 +1753,54 @@ public virtual void TryAttack(Entity target)
...
+                        if (item.BlockAmount < 100)
+                        {
+                            baseDamage -= (int)Math.Round(baseDamage * blockAmount);
+                        }
+                        else
+                        {
+                            baseDamage = 0;
+                        }
+
+                        int absorptionAmount = (int)Math.Round(baseDamage * blockAbsorption);
```
Comment (Arufonsu) [inline Intersect.Server/Entities/Entity.cs:1782]: use `var`

**Review by Arufonsu — CHANGES_REQUESTED** [review]

**Review by Arufonsu — APPROVED** [review]

File: Intersect.Server/Entities/Entity.cs [hunk 8]
```diff
@@ -1753,6 +1753,54 @@ public virtual void TryAttack(Entity target)
...
+
+                if (item != null)
+                {
+                    var originalBaseDamage = baseDamage;
+                    var blockChance = item.BlockChance;
+                    var blockAmount = item.BlockAmount / 100.0;
+                    var blockAbsorption = item.BlockAbsorption / 100.0;
+
+                    //Generate a new attempt to block
+                    if (Randomization.Next(1, 101) < blockChance)
```
Comment (Arufonsu) [inline Intersect.Server/Entities/Entity.cs:1771]: what's the main goal of this randomization (?)
if you want to randomize from 1% to 100% then the logic is fine.
but if u want to rand from 0% to 100% (assuming that you are somewhat simulating percentage this way)
then it's 0,101

**Review by Arufonsu — APPROVED** [review]

**Review by lodicolo — APPROVED** [review]

**Pull request merged**